proto/build.rs

```diff
@@ -40,3 +50,40 @@ fn main() -> Result<()> {
 
     Ok(())
 }
+
+// WARNING: any type attributes adding SERDE_AS **MUST** be placed **BEFORE**
+// SERIALIZE. Otherwise, serde_as is applied to the output of the derive macro
+// (unstable), rather than the other way around.
+// This is a moot point for now, since we can't use field attributes for macro reasons
+// static SERDE_AS: &str = r#"#[::serde_with::serde_as]"#;
+static SERIALIZE: &str = r#"#[derive(::serde::Deserialize, ::serde::Serialize)]"#;
+
+// Requires SERDE_AS on the container
+// :(
+// error: expected non-macro attribute, found attribute macro `::serde_with::serde_as`
+// use tendermint-rs approach instead?
+// https://github.com/penumbra-zone/tendermint-rs/blob/master/proto/src/serializers/bytes.rs#L4-L30
+// static AS_HEX: &str = r#"#[::serde_with::serde_as(as = "::serde_with::hex::Hex")]"#;
+
+static AS_HEX: &str = r#"#[serde(with = "crate::serializers::hexstr")]"#;
+static AS_BASE64: &str = r#"#[serde(with = "crate::serializers::base64str")]"#;
+
+static TYPE_ATTRIBUTES: &[(&str, &str)] = &[
+    //(".penumbra.stake.Validator", SERDE_AS),
+    (".penumbra.stake.Validator", SERIALIZE),
+    (".penumbra.stake.FundingStream", SERIALIZE),
+    //(".penumbra.stake.ValidatorDefinition", SERDE_AS),
+    (".penumbra.stake.ValidatorDefinition", SERIALIZE),
+    (".penumbra.stake.RateData", SERIALIZE),
+    (".penumbra.stake.BaseRateData", SERIALIZE),
+];
+
+static FIELD_ATTRIBUTES: &[(&str, &str)] = &[
+    // TODO: use bech32 here?
+    (".penumbra.stake.Validator.identity_key", AS_HEX),
+    // Using base64 for the validator's consensus key means that
+    // the format is the same as the Tendermint json config files.
+    (".penumbra.stake.Validator.consensus_key", AS_BASE64),
+    (".penumbra.stake.ValidatorDefinition.auth_sig", AS_HEX),
+    (".penumbra.stake.RateData.identity_key", AS_HEX),
+];
```

Comment:
The upshot of this infrastructure is that all of our serialization formats can be determined entirely by the `.proto` files in the crate, together with these "shaping" annotations.  This way, we can confidently rearrange the internals of our domain types, without worrying that by doing so we've accidentally broken the JSON encoding, or something.